<a href="https://colab.research.google.com/github/opoluliakh/data-analytics-learning/blob/main/Poluliakh_HW_%D0%A2%D0%B5%D1%81%D1%82%D0%B8_%D0%B4%D0%BB%D1%8F_%D0%BC%D0%B0%D0%BB%D0%B8%D1%85_%D0%B2%D0%B8%D0%B1%D1%96%D1%80%D0%BE%D0%BA_%D1%82%D0%B0_%D0%BF%D1%80%D0%BE%D0%BF%D0%BE%D1%80%D1%86%D1%96%D0%B9.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Тести для малих вибірок та пропорцій



**Завдання 1**. E-commerce компанія після редизайну сайту підозрює, що **середній час до покупки (time-to-purchase)** користувачів **збільшився**.
Історично середній час від першого заходу на сайт до покупки становив $\mu_0 = 123.7$ хвилин.

Команда аналітиків випадково обрала дані **7 покупців після редизайну**:
`128, 135, 121, 142, 126, 151, 123`

З допомогою наявних даних зʼясуйте, чи збільшився середній час покупки після редизайну.

Для цього виконайте наступні 6 кроків. Правильне виконання кожного з кроків оцінюється в 1 бал.

1. Запишіть нульову та альтернативну гіпотези і визначте тип тесту.
2. Обчисліть вибіркові статистики: $\bar x$, $s$, $n$.
3. Оберіть тип тесту та виконайте його (знайдіть критичне значення тесту, статистику тесту та р-значення) будь-яким пасуючим способом, щоб перевірити гіпотезу на рівні значущості $\alpha = 0.10$.
4. Прийміть рішення, чи відхиляєте ви гіпотезу $H_0$ використовуючи p-value.
5. Напишіть висновок: чи справді редизайн сайту подовжив час до покупки?
6. Чи зміниться ваше рішення при зміні рівня значущості на $\alpha = 0.05$.


## Завдання 1. Чи збільшився середній час до покупки після редизайну?

**Умова:**  
Історично: $\mu_0 = 123.7$ хв  
Дані після редизайну: [128, 135, 121, 142, 126, 151, 123]  
$n = 7$ (мала вибірка)  
Рівень значущості: $\alpha = 0.10$  


### 1. Формулювання гіпотез
$H_0: \mu = 123.7$  
$H_a: \mu > 123.7$  
Тип тесту: **правосторонній (one-sample t-test)**  



In [1]:
import numpy as np
from scipy import stats

sample = np.array([128, 135, 121, 142, 126, 151, 123])

mu0 = 123.7
alpha1 = 0.10
alpha2 = 0.05

x_bar = np.mean(sample)
s = np.std(sample, ddof=1)
n = len(sample)

(x_bar, s, n)


(np.float64(132.28571428571428), np.float64(10.980502200328935), 7)

In [2]:
# t-статистика вручну
t_stat = (x_bar - mu0) / (s / np.sqrt(n))

# критичне значення для правостороннього тесту (df = n-1)
t_crit_1 = stats.t.ppf(1 - alpha1, df=n-1)
t_crit_2 = stats.t.ppf(1 - alpha2, df=n-1)

# p-value
p_value = 1 - stats.t.cdf(t_stat, df=n-1)

t_stat, t_crit_1, t_crit_2, p_value


(np.float64(2.0687273144186475),
 np.float64(1.4397557472577693),
 np.float64(1.9431802805153022),
 np.float64(0.042017654253682646))

In [3]:
for alpha, t_crit in [(alpha1, t_crit_1), (alpha2, t_crit_2)]:
    if (t_stat >= t_crit) and (p_value < alpha):
        print(f"α={alpha}: Відхиляємо H0 → час до покупки зріс.")
    else:
        print(f"α={alpha}: Не відхиляємо H0 → доказів збільшення недостатньо.")


α=0.1: Відхиляємо H0 → час до покупки зріс.
α=0.05: Відхиляємо H0 → час до покупки зріс.


**Завдання 2.**

До спеціальної рекламної кампанії **23%** дорослих упізнавали логотип компанії. Після завершення кампанії відділ маркетингу провів опитування: з **1200** випадково відібраних дорослих **311** упізнали логотип.

Перевірте на рівні значущості **3%** ($\alpha=0.03$), чи дають ці дані достатні підстави стверджувати, що **тепер більше ніж 23%** дорослих упізнають логотип компанії. Для розвʼязку використовуйте бібліотеку `statsmodels`.

Зробіть висновок, чи зросла впізнаваність логотипу.

Додатково, обчисліть довірчий інтревал на заданому рівні значущості і проінтерпретуйте текстом - як він додатково пояснює прийняте нами рішення?

## Завдання 2. Чи зросла впізнаваність логотипу?

До кампанії: \(p_0 = 0.23\)  
Після кампанії: \(n=1200\), успіхів \(x=311\) → \(\hat p = x/n\)  
Рівень значущості: \(\alpha = 0.03\)  
Гіпотези:  
  - \(H_0: p = 0.23\)  
  - \(H_a: p > 0.23\)  (правосторонній тест)


In [4]:
from statsmodels.stats.proportion import proportions_ztest, proportion_confint

x = 311
n = 1200
p0 = 0.23
alpha = 0.03

# 1) Z-test для пропорції (правосторонній)
z_stat, p_value = proportions_ztest(count=x, nobs=n, value=p0, alternative="larger")
p_hat = x / n

print(f"p̂ = {p_hat:.6f}")
print(f"Z = {z_stat:.4f}")
print(f"p-value = {p_value:.5f}")

# 2) Довірчий інтервал для p на рівні 1-α = 97%  (візьмемо Wilson — стійкіший)
ci_low, ci_high = proportion_confint(count=x, nobs=n, alpha=alpha, method="wilson")
print(f"97% CI (Wilson): [{ci_low:.4f}, {ci_high:.4f}]")


p̂ = 0.259167
Z = 2.3058
p-value = 0.01056
97% CI (Wilson): [0.2327, 0.2875]


### Висновок
На рівні значущості α = 0.03 ми отримали p-value = 0.0106 < 0.03, тому відхиляємо нульову гіпотезу.  
Маємо достатньо доказів, що після рекламної кампанії впізнаваність логотипу ** зросла** (частка > 23%).  
Додатково 97% довірчий інтервал [0.233; 0.288] підтверджує висновок, адже повністю лежить вище 0.23.
